### PR TITLE
Feature/server side system report

### PR DESCRIPTION
**Pull request by kvahed**

### CODE REVIEW

File: arangod/RestHandler/RestSystemReportHandler.cpp [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  }
+
+  static std::unordered_map<std::string, std::string> const cmds {
+    {"date", "date -u \"+%Y-%m-%d %H:%M:%S %Z\""},
+    {"dmesg", "dmesg"},
+    {"df", "df -h"},
+    {"memory", "cat /proc/meminfo"},
+    {"uptime", "uptime"},
+    {"uname", "uname -a"},
+    {"top", std::string("top -b -n 1 -H -p ") +
```
Comment (dothebart) [inline arangod/RestHandler/RestSystemReportHandler.cpp]: won't `ProcessInfo TRI_ProcessInfo(TRI_pid_t pid);` also give you these informations?

File: arangod/RestHandler/RestSystemReportHandler.cpp [hunk 2]
```diff
@@ -48,14 +49,17 @@ RestSystemReportHandler::RestSystemReportHandler(
...
-  static std::string const pipe(" | tee ");
-  static std::string const terminate(" > /dev/null");
-  std::system ((cmd + pipe + tmpFile + terminate).c_str());
-  std::ifstream tmpStream(tmpFile.c_str());
-  return std::string(
-    std::istreambuf_iterator<char>(tmpStream), std::istreambuf_iterator<char>());
+std::string exec(std::string const& cmd) {
+  std::array<char, 128> buffer;
+  std::string result;
+  std::unique_ptr<FILE, decltype(&pclose)> pipe(popen(cmd.c_str(), "r"), pclose);
```
Comment (jsteemann) [inline arangod/RestHandler/RestSystemReportHandler.cpp]: The man pages for my system say that `popen` returns a `FILE*`:
```
FILE *popen(const char *command, const char *type);
```
However, they also say that `pipe` expects an `int[2]` as its input parameter, but no `FILE*`:
```
int pipe(int pipefd[2]);
```
How is this supposed to work?

File: arangod/RestHandler/RestSystemReportHandler.cpp [hunk 3]
```diff
@@ -79,26 +99,47 @@ RestStatus RestSystemReportHandler::execute() {
...
+        return RestStatus::DONE;
       }
-    }}
-
+      
+      if (!server.isStopping()) {
+        // Allow only one simultaneous call
+        std::unique_lock<std::mutex> exclusive(_exclusive, std::defer_lock);
+        if (exclusive.try_lock()) {
+          for (auto cmd : cmds) {
```
Comment (jsteemann) [inline arangod/RestHandler/RestSystemReportHandler.cpp]: Still we should not copy the commands all the time. Please use `const&` here.

File: arangod/RestHandler/RestSystemReportHandler.cpp (hunk 2 above)
Comment (ObiWahn) [reply arangod/RestHandler/RestSystemReportHandler.cpp]: `pipe` is a variable name?

Comment (goedderz) [reply arangod/RestHandler/RestSystemReportHandler.cpp]: A small reformulation could make that more clear:
```suggestion
  auto pipe = std::make_unique<FILE, decltype(&pclose)>(popen(cmd.c_str(), "r"), pclose);
```

File: arangod/RestHandler/RestSystemReportHandler.cpp [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+std::unordered_map<std::string, std::string> const cmds {
+  {"date", "time date -u \"+%Y-%m-%d %H:%M:%S %Z\" 2>&1"},
+  {"dmesg", "time dmesg 2>&1"},
+  {"df", "time df -h 2>&1"},
+  {"memory", "time cat /proc/meminfo 2>&1"},
+  {"uptime", "time uptime 2>&1"},
+  {"uname", "time uname -a 2>&1"},
+  {"topp", std::string("time top -b -n 1 -H -p ") +
+      std::to_string(Thread::currentProcessId()) + " 2>&1"},
+  {"top", "time top -b -n 1 2>&1"}
```
Comment (dothebart) [inline arangod/RestHandler/RestSystemReportHandler.cpp]: `lib/Basics/process-utils.h` has `ProcessInfo TRI_ProcessInfo(TRI_pid_t pid);` which offers a portable way to get process metrics.

File: arangod/RestHandler/RestSystemReportHandler.cpp [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+bool RestSystemReportHandler::isAdminUser() const {
+  if (!ExecContext::isAuthEnabled()) {
+    return true;
+  } else {
+    return ExecContext::current().isAdminUser();
+  }
+  return false;
+}
+
+std::unordered_map<std::string, std::string> const cmds {
```
Comment (goedderz) [inline arangod/RestHandler/RestSystemReportHandler.cpp]: Please either put this definition in an anonymous `namespace {}`, or put it as a static member in the `RestSystemReportHandler`.

File: arangod/RestHandler/RestSystemReportHandler.cpp [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+  {"date", "time date -u \"+%Y-%m-%d %H:%M:%S %Z\" 2>&1"},
+  {"dmesg", "time dmesg 2>&1"},
+  {"df", "time df -h 2>&1"},
+  {"memory", "time cat /proc/meminfo 2>&1"},
+  {"uptime", "time uptime 2>&1"},
+  {"uname", "time uname -a 2>&1"},
+  {"topp", std::string("time top -b -n 1 -H -p ") +
+      std::to_string(Thread::currentProcessId()) + " 2>&1"},
```
Comment (goedderz) [inline arangod/RestHandler/RestSystemReportHandler.cpp]: Not to be picky, but if we'd e.g. decide to use `fork` to put `arangod` in the background, this PID would be wrong. Perhaps instead initialize the map lazily in the constructor of the `RestSystemReportHandler`.

File: arangod/RestHandler/RestSystemReportHandler.cpp [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+////////////////////////////////////////////////////////////////////////////////
+/// @brief ArangoDB server
+////////////////////////////////////////////////////////////////////////////////
+
+RestSystemReportHandler::RestSystemReportHandler(
+  application_features::ApplicationServer& server, GeneralRequest* request,
+  GeneralResponse* response) : RestBaseHandler(server, request, response) {}
+
+std::string executeOSCmd(std::string const& cmd) {
+  static std::string const tmpFile("/tmp/arango-inspect.tmp");
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/RestHandler/RestSystemReportHandler.cpp]: This does not have to work on windows yet. It does not break windows.

File: arangod/RestHandler/RestSystemReportHandler.cpp (hunk 1 above)
Comment (kvahed) [reply arangod/RestHandler/RestSystemReportHandler.cpp]: Not all of them, but looking at it and will consider changing down the road.

File: arangod/RestHandler/RestSystemReportHandler.cpp [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+  using namespace std::chrono;
+  auto start = steady_clock::now();
+  
+  VPackBuilder result;
+  { VPackObjectBuilder o(&result);
+    
+    while (true) {
+      
+      if (steady_clock::now()-start > seconds(60)) {
```
Comment (goedderz) [inline arangod/RestHandler/RestSystemReportHandler.cpp]:
```suggestion
      if (steady_clock::now() - start > seconds(60)) {
```

File: arangod/RestHandler/RestSystemReportHandler.h [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+  RestSystemReportHandler(
+    application_features::ApplicationServer&, GeneralRequest*,
+    GeneralResponse*);
+  char const* name() const override final { return "RestSystemReportHandler"; }
+  RequestLane lane() const override final { return RequestLane::CLIENT_SLOW; }
+  RestStatus execute() override;
+private:
+  bool isAdminUser() const;
+
+  std::mutex _exclusive;
```
Comment (goedderz) [inline arangod/RestHandler/RestSystemReportHandler.h]: Shouldn't the `mutex` be static?

File: arangod/RestHandler/RestSystemReportHandler.cpp (hunk 5 above)
Comment (kvahed) [reply arangod/RestHandler/RestSystemReportHandler.cpp]: Very good point, thx

File: arangod/RestHandler/RestSystemReportHandler.cpp (hunk 6 above)
Comment (kvahed) [reply arangod/RestHandler/RestSystemReportHandler.cpp]: Will do thx.

File: arangod/RestHandler/RestSystemReportHandler.h (hunk 9 above)
Comment (kvahed) [reply arangod/RestHandler/RestSystemReportHandler.h]: correct, thx

**Comment by jsteemann** [comment]
Ok, LGTM. Obviously we will need a CHANGELOG entry plus documentation in the `docs` repository, potentially also a description of the HTTP API using DocuBlocks

File: CHANGELOG [hunk 10]
```diff
@@ -80,6 +80,8 @@ devel
 * Fixed a bug in smart graph bfs traversals that might violate path uniqueness
   requirements in rare cases.
 
+* Add acquisition of system report to arangod instances
```
Comment (jsteemann) [inline CHANGELOG]:
```suggestion
* Add acquisition of system report to arangod instances.
```

**Comment by jsteemann** [comment]
PR did not build on any platform after latest merging-in of devel:
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/8139/

**Comment by kvahed** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/8175/

**Comment by jsteemann** [comment]
Build still failed on Linux and Windows.

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/8199/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/8221/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/8349/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/8365/

**Pull request merged**